File: src/main/java/me/centralhardware/znatoki/telegram/statistic/eav/types/Photo.java
package me.centralhardware.znatoki.telegram.statistic.eav.types;

import io.vavr.control.Validation;
import me.centralhardware.znatoki.telegram.statistic.minio.Minio;
import me.centralhardware.znatoki.telegram.statistic.telegram.TelegramSender;
import me.centralhardware.znatoki.telegram.statistic.utils.SpringBootUtils;
import org.telegram.telegrambots.meta.api.methods.GetFile;
import org.telegram.telegrambots.meta.api.objects.PhotoSize;
import org.telegram.telegrambots.meta.api.objects.Update;

import java.time.LocalDateTime;
import java.util.Comparator;
import java.util.Optional;

public non-sealed class Photo implements Type {
    @Override
    public String format(String name, Boolean isOptional) {
        return STR."Отправьте фото \{name}. \{isOptional? optionalText: ""}";
    }

    @Override
    public Validation<String, Void> __validate(Update update, String...variants) {
        return update.hasMessage() && update.getMessage().hasPhoto() ?
                Validation.valid(null) :
                Validation.invalid("Отправьте фото");
    }

    @Override
    public Optional<String> extract(Update update) {
        var sender = SpringBootUtils.getBean(TelegramSender.class);
        var minio = SpringBootUtils.getBean(Minio.class);

        return update.getMessage().getPhoto()
                .stream()
                .max(Comparator.comparing(PhotoSize::getFileSize))
                .map(it -> GetFile.builder()
                        .fileId(it.getFileId())
                        .build())
                .flatMap(sender::downloadFile)
                .flatMap(it -> minio.upload(it, LocalDateTime.now()));
    }

}

File: src/main/kotlin/me/centralhardware/znatoki/telegram/statistic/eav/types/Photo.kt
package me.centralhardware.znatoki.telegram.statistic.eav.types

import arrow.core.Either
import dev.inmo.tgbotapi.AppConfig
import dev.inmo.tgbotapi.extensions.api.files.downloadFile
import dev.inmo.tgbotapi.extensions.api.telegramBot
import dev.inmo.tgbotapi.extensions.utils.asPhotoContent
import dev.inmo.tgbotapi.types.message.abstracts.CommonMessage
import dev.inmo.tgbotapi.types.message.content.MessageContent
import dev.inmo.tgbotapi.types.message.content.PhotoContent
import me.centralhardware.znatoki.telegram.statistic.eav.types.Type.Companion.OPTIONAL_TEXT
import me.centralhardware.znatoki.telegram.statistic.service.MinioService
import java.nio.file.Files
import java.time.LocalDateTime
import kotlin.io.path.writeBytes

data object Photo : Type {
    override fun format(name: String, isOptional: Boolean): String {
        return "Отправьте фото $name. ${if (isOptional) OPTIONAL_TEXT else ""}"
    }

    override fun validate(
        message: CommonMessage<MessageContent>,
        variants: List<String>,
    ): Either<String, Unit> {
        return if (message.content is PhotoContent) {
            Either.Right(Unit)
        } else {
            Either.Left("Отправьте фото")
        }
    }

    override suspend fun extract(message: CommonMessage<MessageContent>): String {
        val photo = message.content.asPhotoContent()!!.media
        val temp = Files.createTempFile("tg", "sdf")
        temp.writeBytes(telegramBot(AppConfig.botToken()).downloadFile(photo))
        return MinioService.upload(temp.toFile(), LocalDateTime.now()).getOrThrow()
    }
}
